protocols/directfund/directfund.go

```diff
@@ -358,6 +359,51 @@ func ConstructObjectiveFromMessage(m protocols.Message, myAddress types.Address)
 	return objective, nil
 }
 
+// ObjectiveRequest represents a request to create a new direct funding objective.
+type ObjectiveRequest struct {
+	MyAddress         types.Address
+	CounterParty      types.Address
+	AppDefinition     types.Address
+	AppData           types.Bytes
+	ChallengeDuration *types.Uint256
+	Outcome           outcome.Exit
+	Nonce             int64
+}
+
+// Id returns the objective id for the request.
+func (r ObjectiveRequest) Id() protocols.ObjectiveId {
+	fixedPart := state.FixedPart{ChainId: big.NewInt(0), // TODO
+		Participants:      []types.Address{r.MyAddress, r.CounterParty},
+		ChannelNonce:      big.NewInt(r.Nonce),
+		ChallengeDuration: r.ChallengeDuration}
+
+	channelId, _ := fixedPart.ChannelId()
+	return protocols.ObjectiveId(ObjectivePrefix + channelId.String())
+}
+
+// SpawnObjective creates a new direct funding objective from a given request.
+func SpawnObjective(request ObjectiveRequest) (Objective, error) {
```

Comment:
It's not completely clear to me why we need to have both a `NewObjective` function and a `SpawnObjective` function. It seems like `SpawnObjective` is just another constructor for an `Objective` -- can it be removed?

Reply:
`SpawnObjective` and `NewObjective` could probably get merged into one constructor that takes in an `ObjectiveRequest`. We'd probably need to update some tests that reference `NewObjective` but it wouldn't be too much work.

Reply:
I've refactored this so `NewObjective` now takes a `ObjectiveRequest`. The previous constructor is now a private function `constructFromInitialState`